specs/TransportProtocols.md

```diff
@@ -0,0 +1,176 @@
+
+## Transport Requirements
+
+A transport is required to have the following attributes:
+
+1. Duplex - Able to send messages from Server to Client and from Client to Server
+1. Frame-based - Messages have fixed length (as opposed to streaming, where data is just pushed throughout the connection)
+1. Frame Metadata - Able to encode two pieces of Frame Metadata
+    * `Type` - Either `Binary` or `Text`.
+    * `EndOfMessage` - Boolean indicating if this frame represents the final frame of a complete message.
+1. Binary-safe - Able to transmit arbitrary binary data, regardless of content
+1. Text-safe - Able to transmit arbitrary text data, preserving the content. Line-endings must be preserved **but may be converted to a different format**. For example `\r\n` may be converted to `\n`. This is due to quirks in some transports (Server Sent Events)
+
+The only transport which fully implements the duplex requirement is WebSockets, the others are "half-transports" which implement one end of the duplex connection. They are used in combination to achieve a duplex connection.
+
+Throughout this document, the term `[endpoint-base]` is used to refer to the route assigned to a particular end point
+
+## WebSockets (Full Duplex)
+
+The WebSockets transport is unique in that it is full duplex, and a persistent connection that can be established in a single operation. As a result, the client is not required to use the `[endpoint-base]/negotiate` endpoint to establish a connection in advance. It also includes all the necessary metadata in it's own frame metadata.
+
+The WebSocket transport is activated by making a WebSocket connection to `[endpoint-base]/ws`. Upon doing so, the connection is fully established and immediately ready for frames to be sent/received. The WebSocket OpCode field is used to indicate the type of the frame (Text or Binary) and the WebSocket "FIN" flag is used to indicate the end of a message.
+
+Establishing a second WebSocket connection when there is already a WebSocket connection associated with the Endpoints connection is not permitted and will fail with a `400` (Bad Request) status code.
+
+## HTTP Post (Client-to-Server only)
+
+HTTP Post is a half-transport, it is only able to send messages from the Client to the Server, as such it is always used with one of the other half-transports which can send from Server to Client (Server Sent Events and Long Polling).
+
+This transport uses HTTP Post requests to a specific end point to encode messages. Each HTTP Request represents a single message (this achieves the Frame-based requirement), contains raw binary data (Binary-safe) and uses the query string to encode the metadata.
+
+This transport requires that a connection be established using the `[endpoint-base]/negotiate` end point.
+
+The HTTP request is made to the URL `[endpoint-base]/send` with the following query string parameters
```

Comment:
RESTy just meaning using the path instead of query-string? Because that's not actually how REST works... Just moving things to the path does not a REST API make.